# Dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

In [1]:
import numpy as np

In [2]:
import pandas as pd

In [3]:
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

The dataset has the classified entries stored mainly as object with some integers. There are 50,000 rows, some having null objects.

In [6]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',  'vehicle_type', 'registration_year', 'gear_box',  'power_ps', 'model', 'odometer', 'registration_month', 'fuel_type', 'brand', 'unrepaired_damage', 'ad_created', 'no_pictures', 'postal_code', 'last_seen'] 

In [7]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-09 11:54:38  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year gear_box      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  no_pictures   postal_code            last_seen  
count                 50000      50000.0  50000.000000                50000  
unique                   76          NaN           NaN                39481  
top     2016-04-03 00:00:00          NaN           NaN  2016-04-07 06:17:27  
freq                   1946          NaN           NaN                    8  
mean                    NaN          0.0  50813.627300                  NaN  
std                     NaN          0.0  25779.747957                  NaN  
min                     NaN          0.0   1067.000000                  NaN  
25%                     NaN          0.0  30451.000000                  NaN  
50%                     NaN          0.0  49577.000000                  NaN  
75%                     NaN          0.0  71540.000000                  NaN  
max                     NaN          0.0  99998.000000                  NaN

## Columns such as price and odometer are not numbers & need to be converted 

In [8]:
autos['price'].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

In [9]:
autos['odometer'].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [10]:
autos['price'] = autos['price'].str.replace('$', '')
autos['price'] = autos['price'].str.replace(',', '')

In [11]:
autos['price'] = autos['price'].astype(int)

In [12]:
autos['odometer'] = autos['odometer'].str.replace(",", '')
autos['odometer'] = autos['odometer'].str.replace('km', '')

In [13]:
autos['odometer'] = autos['odometer'].astype(int)

In [14]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
seller                50000 non-null object
offer_type            50000 non-null object
price                 50000 non-null int64
ab_test               50000 non-null object
vehicle_type          44905 non-null object
registration_year     50000 non-null int64
gear_box              47320 non-null object
power_ps              50000 non-null int64
model                 47242 non-null object
odometer              50000 non-null int64
registration_month    50000 non-null int64
fuel_type             45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null object
no_pictures           50000 non-null int64
postal_code           50000 non-null int64
last_seen             50000 non-null object
dtypes: int64(7), 

In [15]:
autos.rename(columns={'odometer': 'odometer_kms'}, inplace=True)

In [16]:
autos['seller'].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

Seller is almost all the same value - can be dropped

In [17]:
autos['offer_type'].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

Offer type can also be dropped

## Looking for outliers in price and odometer values

In [18]:
unique = autos['price']
unique_dict = {}
for i in unique:
    if i in  unique_dict:
        unique_dict[i] += 1
    else:
        unique_dict[i] = 1
        

In [20]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [21]:
autos['price'].sort_values(ascending=False).head(30)

39705    99999999
42221    27322222
39377    12345678
47598    12345678
27371    12345678
2897     11111111
24384    11111111
11137    10000000
47634     3890000
7814      1300000
22947     1234566
43049      999999
514        999999
37585      999990
36818      350000
14715      345000
34723      299000
35923      295000
12682      265000
47337      259000
38299      250000
37840      220000
40918      198000
43668      197000
28090      194000
20351      190000
17140      180000
11433      175000
32840      169999
18509      169000
Name: price, dtype: int64

### Taking outlier prices out - keeping prices between 99 and 350001

In [22]:
autos = autos[autos['price'].between(99,350001)]

In [23]:
autos['price'].describe()

count     48243.000000
mean       5928.074809
std        9077.322281
min          99.000000
25%        1250.000000
50%        3000.000000
75%        7499.000000
max      350000.000000
Name: price, dtype: float64

In [24]:
autos['odometer_kms'].describe()

count     48243.000000
mean     125923.035466
std       39541.689312
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_kms, dtype: float64

In [25]:
autos['odometer_kms'].sort_values(ascending=True)

21421      5000
25837      5000
10789      5000
30675      5000
43032      5000
33304      5000
25903      5000
10648      5000
25916      5000
10628      5000
10621      5000
25925      5000
10607      5000
10584      5000
30670      5000
25936      5000
10541      5000
25964      5000
43233      5000
10456      5000
10442      5000
10429      5000
10368      5000
26024      5000
1011       5000
43342      5000
10249      5000
26103      5000
43422      5000
10556      5000
          ...  
20234    150000
20225    150000
20235    150000
20238    150000
20239    150000
20240    150000
20241    150000
20242    150000
20243    150000
20237    150000
20223    150000
20222    150000
20221    150000
20195    150000
20196    150000
20198    150000
20199    150000
20200    150000
20201    150000
20202    150000
20203    150000
20206    150000
20207    150000
20208    150000
20214    150000
20216    150000
20217    150000
20220    150000
20191    150000
49999    150000
Name: odometer_kms, Leng

There seems to be no odometer outliers needing removal

## Converting and checking dates

In [26]:
autos['date_crawled'] = autos['date_crawled'].str[:10]

In [27]:
autos['date_crawled'].describe()

count          48243
unique            34
top       2016-04-03
freq            1862
Name: date_crawled, dtype: object

In [28]:
autos['date_crawled'].value_counts(normalize=True, dropna=False)

2016-04-03    0.038596
2016-03-20    0.037788
2016-03-21    0.037228
2016-03-12    0.036897
2016-03-14    0.036648
2016-04-04    0.036544
2016-03-07    0.036047
2016-04-02    0.035591
2016-03-28    0.034948
2016-03-19    0.034762
2016-03-15    0.034305
2016-03-29    0.034098
2016-03-30    0.033725
2016-04-01    0.033704
2016-03-08    0.033207
2016-03-09    0.033000
2016-03-22    0.032917
2016-03-11    0.032606
2016-03-26    0.032316
2016-03-23    0.032295
2016-03-10    0.032274
2016-03-31    0.031860
2016-03-17    0.031528
2016-03-25    0.031507
2016-03-27    0.031113
2016-03-24    0.029455
2016-03-16    0.029455
2016-03-05    0.025372
2016-03-13    0.015671
2016-03-06    0.014033
2016-04-05    0.013059
2016-03-18    0.012893
2016-04-06    0.003171
2016-04-07    0.001389
Name: date_crawled, dtype: float64

In [29]:
autos['date_crawled'].sort_index()

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
5        2016-03-21
6        2016-03-20
7        2016-03-16
8        2016-03-22
9        2016-03-16
10       2016-03-15
11       2016-03-16
12       2016-03-31
13       2016-03-23
14       2016-03-23
15       2016-04-01
16       2016-03-16
17       2016-03-29
18       2016-03-26
19       2016-03-17
20       2016-03-05
21       2016-03-06
22       2016-03-28
23       2016-03-10
24       2016-04-03
26       2016-04-03
28       2016-03-19
29       2016-04-02
31       2016-03-14
32       2016-03-20
            ...    
49968    2016-04-01
49969    2016-03-17
49970    2016-03-21
49971    2016-03-29
49972    2016-03-26
49973    2016-03-27
49975    2016-03-27
49976    2016-03-19
49977    2016-03-31
49978    2016-04-04
49979    2016-03-20
49980    2016-03-12
49981    2016-03-15
49982    2016-03-29
49983    2016-03-06
49985    2016-04-02
49986    2016-04-04
49987    2016-03-22
49988    2016-03-28


In [30]:
autos['ad_created'] = autos['ad_created'].str[:10]

In [31]:
autos['ad_created'].describe()

count          48243
unique            76
top       2016-04-03
freq            1874
Name: ad_created, dtype: object

In [32]:
autos['ad_created'].value_counts(normalize=True, dropna=False)

2016-04-03    0.038845
2016-03-20    0.037850
2016-03-21    0.037456
2016-04-04    0.036897
2016-03-12    0.036731
2016-03-14    0.035280
2016-04-02    0.035280
2016-03-28    0.035052
2016-03-07    0.034782
2016-03-29    0.034057
2016-03-15    0.034036
2016-04-01    0.033684
2016-03-19    0.033642
2016-03-30    0.033539
2016-03-08    0.033207
2016-03-09    0.033083
2016-03-11    0.032917
2016-03-22    0.032730
2016-03-26    0.032378
2016-03-23    0.032129
2016-03-10    0.031984
2016-03-31    0.031901
2016-03-25    0.031632
2016-03-17    0.031196
2016-03-27    0.031030
2016-03-16    0.029953
2016-03-24    0.029393
2016-03-05    0.022926
2016-03-13    0.017039
2016-03-06    0.015298
                ...   
2016-02-02    0.000041
2016-02-12    0.000041
2016-02-24    0.000041
2016-02-26    0.000041
2016-02-14    0.000041
2016-02-18    0.000041
2016-02-05    0.000041
2016-01-10    0.000041
2016-02-20    0.000041
2016-02-22    0.000021
2015-12-30    0.000021
2016-01-13    0.000021
2016-02-11 

In [33]:
autos['last_seen'] = autos['last_seen'].str[:10]

In [34]:
autos['last_seen'].describe()

count          48243
unique            34
top       2016-04-06
freq           10707
Name: last_seen, dtype: object

In [35]:
autos['last_seen'].value_counts(normalize=True, dropna=False)

2016-04-06    0.221939
2016-04-07    0.132102
2016-04-05    0.125054
2016-03-17    0.028087
2016-04-03    0.025123
2016-04-02    0.024895
2016-03-30    0.024708
2016-04-04    0.024542
2016-03-31    0.023817
2016-03-12    0.023775
2016-04-01    0.022843
2016-03-29    0.022283
2016-03-22    0.021350
2016-03-28    0.020853
2016-03-20    0.020666
2016-03-21    0.020563
2016-03-24    0.019775
2016-03-25    0.019132
2016-03-23    0.018593
2016-03-26    0.016666
2016-03-16    0.016438
2016-03-15    0.015878
2016-03-19    0.015795
2016-03-27    0.015546
2016-03-14    0.012624
2016-03-11    0.012396
2016-03-10    0.010634
2016-03-09    0.009597
2016-03-13    0.008872
2016-03-18    0.007317
2016-03-08    0.007317
2016-03-07    0.005431
2016-03-06    0.004312
2016-03-05    0.001078
Name: last_seen, dtype: float64

In [36]:
autos['registration_year'].describe()

count    48243.000000
mean      2004.729702
std         87.880224
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

####  From the above there are years included which look incorrect including min of 1000 and 9999

In [37]:
autos['registration_year'].value_counts()

2000    3105
2005    2912
1999    2879
2004    2699
2003    2694
2006    2668
2001    2629
2002    2477
1998    2344
2007    2273
2008    2210
2009    2080
1997    1929
2011    1618
2010    1587
2017    1384
1996    1357
2012    1309
2016    1203
1995    1194
2013     801
2014     662
1994     627
2018     468
1993     420
2015     381
1992     367
1991     338
1990     332
1989     171
        ... 
1934       2
1951       2
1941       2
1954       2
1955       2
2019       2
1910       2
1957       2
1800       2
8888       1
1950       1
4500       1
1929       1
1953       1
1111       1
1001       1
4800       1
1931       1
5911       1
1948       1
1927       1
1943       1
1939       1
1938       1
1000       1
2800       1
4100       1
6200       1
9000       1
1952       1
Name: registration_year, Length: 95, dtype: int64

In [38]:
autos = autos[autos['registration_year'].between(1920,2017)]

In [39]:
autos['registration_year'].value_counts(normalize=True)

2000    0.065023
2005    0.060982
1999    0.060291
2004    0.056521
2003    0.056416
2006    0.055872
2001    0.055055
2002    0.051872
1998    0.049087
2007    0.047600
2008    0.046281
2009    0.043558
1997    0.040396
2011    0.033883
2010    0.033234
2017    0.028983
1996    0.028418
2012    0.027412
2016    0.025193
1995    0.025004
2013    0.016774
2014    0.013863
1994    0.013130
1993    0.008795
2015    0.007979
1992    0.007686
1991    0.007078
1990    0.006953
1989    0.003581
1988    0.002785
          ...   
1977    0.000461
1960    0.000461
1976    0.000440
1969    0.000398
1975    0.000377
1965    0.000356
1964    0.000251
1963    0.000168
1959    0.000126
1961    0.000126
1956    0.000084
1958    0.000084
1962    0.000084
1937    0.000084
1954    0.000042
1941    0.000042
1934    0.000042
1951    0.000042
1957    0.000042
1955    0.000042
1950    0.000021
1953    0.000021
1948    0.000021
1929    0.000021
1931    0.000021
1938    0.000021
1943    0.000021
1939    0.0000

Exploring the brands of cars

In [40]:
autos['brand'].value_counts(normalize=True).head(20)

volkswagen        0.212787
bmw               0.108959
opel              0.108058
mercedes_benz     0.096038
audi              0.086237
ford              0.069756
renault           0.047621
peugeot           0.029590
fiat              0.025884
seat              0.018596
skoda             0.016167
nissan            0.015371
mazda             0.015287
smart             0.014324
citroen           0.014136
toyota            0.012753
hyundai           0.009989
sonstige_autos    0.009361
volvo             0.009068
mini              0.008733
Name: brand, dtype: float64

In [41]:
autos['brand'].value_counts().sort_values(ascending=False).head(20).index

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'sonstige_autos', 'volvo', 'mini'],
      dtype='object')

In [42]:
top_brands = autos['brand'].value_counts().sort_values(ascending=False).head(20).index

In [43]:
car_brands = {}
for brand in autos['brand'][1:]:
    if brand in car_brands:
        car_brands[brand] += 1
    else:
        car_brands[brand] = 1

In [44]:
print(car_brands)

{'nissan': 734, 'ford': 3331, 'kia': 340, 'daihatsu': 119, 'honda': 379, 'mazda': 730, 'skoda': 772, 'mitsubishi': 387, 'mercedes_benz': 4586, 'daewoo': 74, 'jeep': 107, 'hyundai': 477, 'alfa_romeo': 318, 'volvo': 433, 'porsche': 279, 'chevrolet': 268, 'lada': 27, 'chrysler': 168, 'suzuki': 280, 'subaru': 99, 'fiat': 1236, 'lancia': 53, 'saab': 78, 'opel': 5160, 'land_rover': 99, 'jaguar': 71, 'trabant': 65, 'volkswagen': 10161, 'peugeot': 1412, 'rover': 65, 'smart': 684, 'toyota': 609, 'mini': 417, 'audi': 4118, 'citroen': 675, 'dacia': 128, 'seat': 888, 'bmw': 5203, 'renault': 2274, 'sonstige_autos': 447}


In [45]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [46]:
mean_price = {}
for car in autos['brand'].unique():
    selected = autos[autos['brand'] == car]
    mean_price[car] = selected['price'].mean()
    
    

In [47]:
print(mean_price)

{'nissan': 4683.061307901908, 'ford': 3761.099069348544, 'kia': 5942.905882352941, 'daihatsu': 1625.1512605042017, 'honda': 4052.651715039578, 'mazda': 4082.6534246575343, 'skoda': 6382.124352331606, 'mitsubishi': 3422.891472868217, 'mercedes_benz': 8572.807239424335, 'daewoo': 1105.6756756756756, 'jeep': 11590.214953271028, 'hyundai': 5391.7945492662475, 'alfa_romeo': 4067.393081761006, 'volvo': 4917.5103926097, 'porsche': 46788.444444444445, 'chevrolet': 6762.462686567164, 'lada': 2688.296296296296, 'chrysler': 3539.9166666666665, 'suzuki': 4157.9, 'subaru': 4001.090909090909, 'fiat': 2804.398867313916, 'lancia': 3230.169811320755, 'saab': 3192.269230769231, 'opel': 2980.7288759689923, 'land_rover': 18934.272727272728, 'jaguar': 11961.56338028169, 'trabant': 1843.5384615384614, 'volkswagen': 5382.330184037004, 'peugeot': 3090.8648266100495, 'rover': 1586.4923076923078, 'smart': 3550.3055555555557, 'toyota': 5151.940886699507, 'mini': 10566.824940047962, 'audi': 9286.365225837786, 'ci

### Average price of all cars

In [48]:
autos['price'].mean()

5946.908862456023

In [49]:
sorted(mean_price.items(), key=lambda x: x[1], reverse=True)

[('porsche', 46788.444444444445),
 ('land_rover', 18934.272727272728),
 ('sonstige_autos', 12725.787472035794),
 ('jaguar', 11961.56338028169),
 ('jeep', 11590.214953271028),
 ('mini', 10566.824940047962),
 ('audi', 9286.365225837786),
 ('mercedes_benz', 8572.807239424335),
 ('bmw', 8328.581971939266),
 ('chevrolet', 6762.462686567164),
 ('skoda', 6382.124352331606),
 ('kia', 5942.905882352941),
 ('dacia', 5920.3828125),
 ('hyundai', 5391.7945492662475),
 ('volkswagen', 5382.330184037004),
 ('toyota', 5151.940886699507),
 ('volvo', 4917.5103926097),
 ('nissan', 4683.061307901908),
 ('seat', 4388.873873873874),
 ('suzuki', 4157.9),
 ('mazda', 4082.6534246575343),
 ('alfa_romeo', 4067.393081761006),
 ('honda', 4052.651715039578),
 ('subaru', 4001.090909090909),
 ('citroen', 3769.6637037037035),
 ('ford', 3761.099069348544),
 ('smart', 3550.3055555555557),
 ('chrysler', 3539.9166666666665),
 ('mitsubishi', 3422.891472868217),
 ('lancia', 3230.169811320755),
 ('saab', 3192.269230769231),
 

### Finding the top brands based on the numbers of cars sold

In [50]:
top_20 = autos['brand'].value_counts().sort_values(ascending=False).head(20).index

In [51]:
print(top_20)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'sonstige_autos', 'volvo', 'mini'],
      dtype='object')


### Finding the mean prices of the top 20 brands sold

In [52]:
top_mean_price = {}
for car in top_20:
    selected = autos[autos['brand'] == car]
    top_mean_price[car] = selected['price'].mean()

In [53]:
print(top_mean_price)

{'nissan': 4683.061307901908, 'fiat': 2804.398867313916, 'opel': 2980.7288759689923, 'volvo': 4917.5103926097, 'mazda': 4082.6534246575343, 'audi': 9286.365225837786, 'mercedes_benz': 8572.807239424335, 'volkswagen': 5382.330184037004, 'peugeot': 3090.8648266100495, 'smart': 3550.3055555555557, 'hyundai': 5391.7945492662475, 'toyota': 5151.940886699507, 'mini': 10566.824940047962, 'skoda': 6382.124352331606, 'ford': 3761.099069348544, 'citroen': 3769.6637037037035, 'seat': 4388.873873873874, 'bmw': 8328.581971939266, 'renault': 2461.5967458223395, 'sonstige_autos': 12725.787472035794}


In [55]:
sorted(top_mean_price.items(), key=lambda x: x[1], reverse=True)

[('sonstige_autos', 12725.787472035794),
 ('mini', 10566.824940047962),
 ('audi', 9286.365225837786),
 ('mercedes_benz', 8572.807239424335),
 ('bmw', 8328.581971939266),
 ('skoda', 6382.124352331606),
 ('hyundai', 5391.7945492662475),
 ('volkswagen', 5382.330184037004),
 ('toyota', 5151.940886699507),
 ('volvo', 4917.5103926097),
 ('nissan', 4683.061307901908),
 ('seat', 4388.873873873874),
 ('mazda', 4082.6534246575343),
 ('citroen', 3769.6637037037035),
 ('ford', 3761.099069348544),
 ('smart', 3550.3055555555557),
 ('peugeot', 3090.8648266100495),
 ('opel', 2980.7288759689923),
 ('fiat', 2804.398867313916),
 ('renault', 2461.5967458223395)]

### Top 5 brands and mean prices 

|  Brand  |  Mean Price  |
|------|------|
| Volkswagen | 5382   |
    |  bmw   | 8328  |
    |   opel  |   2980 |
    | mercedes_benz | 8572  |
| audi   | 9286  |

#### With the mean price across the data being $5946, from the above, the top 5 brands include 3 premium brands - BMW, MB, Audi - one average brand - VW - and one cheap brand - Opel

Use the loop method to calculate the mean mileage and mean price for each of the top brands, storing the results in a dictionary.

In [57]:
top_mean_milage = {}
for car in top_20:
    selected = autos[autos['brand'] == car]
    top_mean_milage[car] = selected['odometer_kms'].mean()

In [58]:
print(top_mean_milage)

{'nissan': 118678.47411444141, 'fiat': 117269.41747572816, 'opel': 129487.40310077519, 'volvo': 138787.52886836027, 'mazda': 124753.42465753424, 'audi': 129519.18406993686, 'mercedes_benz': 131196.03139991278, 'volkswagen': 129022.24190532428, 'peugeot': 127275.30077848549, 'smart': 100562.86549707603, 'hyundai': 107148.84696016772, 'toyota': 116338.25944170772, 'mini': 89100.71942446043, 'skoda': 111042.74611398963, 'ford': 124265.98619033323, 'citroen': 119933.33333333333, 'seat': 122015.76576576577, 'bmw': 132766.673073227, 'renault': 128370.71240105541, 'sonstige_autos': 90738.25503355704}


In [59]:
bmp_series = pd.Series(top_mean_price)

In [60]:
print(bmp_series)

audi               9286.365226
bmw                8328.581972
citroen            3769.663704
fiat               2804.398867
ford               3761.099069
hyundai            5391.794549
mazda              4082.653425
mercedes_benz      8572.807239
mini              10566.824940
nissan             4683.061308
opel               2980.728876
peugeot            3090.864827
renault            2461.596746
seat               4388.873874
skoda              6382.124352
smart              3550.305556
sonstige_autos    12725.787472
toyota             5151.940887
volkswagen         5382.330184
volvo              4917.510393
dtype: float64


In [61]:
bmm_series = pd.Series(top_mean_milage)

In [62]:
print(bmm_series)

audi              129519.184070
bmw               132766.673073
citroen           119933.333333
fiat              117269.417476
ford              124265.986190
hyundai           107148.846960
mazda             124753.424658
mercedes_benz     131196.031400
mini               89100.719424
nissan            118678.474114
opel              129487.403101
peugeot           127275.300778
renault           128370.712401
seat              122015.765766
skoda             111042.746114
smart             100562.865497
sonstige_autos     90738.255034
toyota            116338.259442
volkswagen        129022.241905
volvo             138787.528868
dtype: float64


In [69]:
df = pd.DataFrame(round(bmp_series), columns=['mean_price'])


In [73]:
df.insert(1, 'mean_mileage', bmm_series)

In [79]:
print(df.round(1))

                mean_price  mean_mileage
audi                9286.4      129519.2
bmw                 8328.6      132766.7
citroen             3769.7      119933.3
fiat                2804.4      117269.4
ford                3761.1      124266.0
hyundai             5391.8      107148.8
mazda               4082.7      124753.4
mercedes_benz       8572.8      131196.0
mini               10566.8       89100.7
nissan              4683.1      118678.5
opel                2980.7      129487.4
peugeot             3090.9      127275.3
renault             2461.6      128370.7
seat                4388.9      122015.8
skoda               6382.1      111042.7
smart               3550.3      100562.9
sonstige_autos     12725.8       90738.3
toyota              5151.9      116338.3
volkswagen          5382.3      129022.2
volvo               4917.5      138787.5


### There seems to be a link between lower milage and higher prices - looking at Mini and Sonstiage Autos